In [ ]:
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
heart_disease = fetch_ucirepo(id=45) 
  
# data (as pandas dataframes) 
X = heart_disease.data.features 
y = heart_disease.data.targets 
  
# metadata 
print(heart_disease.metadata) 
  
# variable information 
print(heart_disease.variables) 

# 데이터 가져오기


In [ ]:
import pandas as pd
import numpy as np

# 1. Feature(X)와 Target(y)을 하나의 데이터프레임으로 병합
X = heart_disease.data.features 
y = heart_disease.data.targets 
df = pd.concat([X, y], axis=1)

# 2. 타겟 변수(num)를 이진 분류(target)로 변환 (0: 정상, 1~4: 심장병 있음 -> 1)
df['target'] = df['num'].apply(lambda x: 1 if x > 0 else 0)
df = df.drop(columns=['num']) # 기존 num 컬럼은 제거

print("=== 1. 데이터셋 변환 후 상위 5개 행 ===")
print(df.head())

print("\n=== 2. 실제 결측치가 있는 컬럼과 개수 ===")
missing_info = df.isnull().sum()
print(missing_info[missing_info > 0])

print("\n=== 3. 클래스 불균형 확인 (정상 vs 환자) ===")
print(df['target'].value_counts(normalize=True))

# 데이터 불러온 결과 확인

=== 1. 데이터셋 변환 후 상위 5개 행 ===
   age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   1       145   233    1        2      150      0      2.3      3   
1   67    1   4       160   286    0        2      108      1      1.5      2   
2   67    1   4       120   229    0        2      129      1      2.6      2   
3   37    1   3       130   250    0        0      187      0      3.5      3   
4   41    0   2       130   204    0        2      172      0      1.4      1   

    ca  thal  target  
0  0.0   6.0       0  
1  3.0   3.0       1  
2  2.0   7.0       1  
3  0.0   3.0       0  
4  0.0   3.0       0  

=== 2. 실제 결측치가 있는 컬럼과 개수 ===
ca      4
thal    2
dtype: int64

=== 3. 클래스 불균형 확인 (정상 vs 환자) ===
target
0    0.541254
1    0.458746
Name: proportion, dtype: float64


In [ ]:
import pandas as pd

# 1. 아까 가져온 X(특성)와 y(정답)를 하나의 데이터프레임으로 합침
X = heart_disease.data.features 
y = heart_disease.data.targets 
df = pd.concat([X, y], axis=1)

# 2. [head] 데이터의 처음 5개 행을 시각적으로 확인
print("=================== 1. df.head() ===================")
print("데이터가 어떤 형태로 생겼는지 상위 5개 행을 보여줍니다.")
print("-" * 52)
print(df.head())
print("\n" + "="*52 + "\n")

# 3. [info] 데이터의 변수명, 비결측치 개수, 데이터 타입 확인
print("=================== 2. df.info() ===================")
print("각 컬럼의 데이터 타입(정수, 실수, 문자)과 결측치 여부를 알려줍니다.")
print("-" * 52)
df.info() # info()는 자체적으로 출력을 해주므로 print()로 감싸지 않아도 됩니다.
print("\n" + "="*52 + "\n")

# 4. [describe] 수치형 데이터의 기술 통계량 확인
print("================= 3. df.describe() =================")
print("수치형 변수들의 평균, 표준편차, 최솟값, 사분위수, 최댓값을 보여줍니다.")
print("-" * 52)
print(df.describe())

# head, info 결과 출력

=================== 1. df.head() ===================
데이터가 어떤 형태로 생겼는지 상위 5개 행을 보여줍니다.
----------------------------------------------------
   age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   1       145   233    1        2      150      0      2.3      3   
1   67    1   4       160   286    0        2      108      1      1.5      2   
2   67    1   4       120   229    0        2      129      1      2.6      2   
3   37    1   3       130   250    0        0      187      0      3.5      3   
4   41    0   2       130   204    0        2      172      0      1.4      1   

    ca  thal  num  
0  0.0   6.0    0  
1  3.0   3.0    2  
2  2.0   7.0    1  
3  0.0   3.0    0  
4  0.0   3.0    0  


=================== 2. df.info() ===================
각 컬럼의 데이터 타입(정수, 실수, 문자)과 결측치 여부를 알려줍니다.
----------------------------------------------------
<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column

In [ ]:
print("========= 3. df.describe() 수치형 통계량 =========")
# describe() 결과를 print로 감싸서 강제 출력합니다.
print(df.describe())

# describe 결과 출력

========= 3. df.describe() 수치형 통계량 =========
              age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.438944    0.679868    3.158416  131.689769  246.693069    0.148515   
std      9.038662    0.467299    0.960126   17.599748   51.776918    0.356198   
min     29.000000    0.000000    1.000000   94.000000  126.000000    0.000000   
25%     48.000000    0.000000    3.000000  120.000000  211.000000    0.000000   
50%     56.000000    1.000000    3.000000  130.000000  241.000000    0.000000   
75%     61.000000    1.000000    4.000000  140.000000  275.000000    0.000000   
max     77.000000    1.000000    4.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  299.000000   
mean     0.990099  149.607261    0.326733    1.039604    1.6006

In [7]:
# 1. 타겟 변수 이진 분류 정제 (앞서 변환하지 않았다면 실행)
df['target'] = df['num'].apply(lambda x: 1 if x > 0 else 0)

# 2. 비율 출력 (normalize=True를 주면 비율로 계산됩니다)
print("=== 타겟 클래스 비율 (정상 0, 심장병 1) ===")
print(df['target'].value_counts(normalize=True))

print("\n=== 실제 데이터 개수 ===")
print(df['target'].value_counts())

# 정상(0)과 심장병 환자(1) 비율 백분율로 환산해 확인

=== 타겟 클래스 비율 (정상 0, 심장병 1) ===
target
0    0.541254
1    0.458746
Name: proportion, dtype: float64

=== 실제 데이터 개수 ===
target
0    164
1    139
Name: count, dtype: int64


In [8]:
# 연속형 특성 목록 정의
numeric_features = ['age', 'trestbps', 'chol', 'thalach', 'oldpeak']

print("=== [1] 연속형 특성별 결측치 개수 및 비율 ===")
missing_counts = df[numeric_features].isnull().sum()
missing_ratios = df[numeric_features].isnull().mean() * 100

missing_df = pd.DataFrame({
    'Missing Values (개)': missing_counts,
    'Ratio (%)': missing_ratios
})
print(missing_df)

print("\n=== [2] 전체 데이터셋 기준 총 결측치 개수 ===")
print(f"전체 결측치 수: {df.isnull().sum().sum()}개")

# 결측값 탐지

=== [1] 연속형 특성별 결측치 개수 및 비율 ===
          Missing Values (개)  Ratio (%)
age                        0        0.0
trestbps                   0        0.0
chol                       0        0.0
thalach                    0        0.0
oldpeak                    0        0.0

=== [2] 전체 데이터셋 기준 총 결측치 개수 ===
전체 결측치 수: 6개


In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# 시각화 스타일 설정
sns.set_theme(style="whitegrid")
plt.rcParams['font.family'] = 'Malgun Gothic' # 윈도우 한글 깨짐 방지 (맥은 'AppleGothic')
plt.rcParams['axes.unicode_minus'] = False

# 5개 특성에 대한 Boxplot 그리기
fig, axes = plt.subplots(2, 3, figsize=(15, 10))
axes = axes.flatten() # 2차원 배열을 1차원으로 펼치기

for i, col in enumerate(numeric_features):
    sns.boxplot(data=df, y=col, ax=axes[i], color='skyblue')
    axes[i].set_title(f'{col}의 분포 및 이상치', fontsize=12)
    axes[i].set_ylabel('')

# 남는 6번째 서브플롯 비우기
axes[-1].axis('off')

plt.tight_layout()
plt.show()

# 이상치 탐지

In [10]:
print("=== IQR 기반 수치형 특성별 이상치 탐지 ===")
print("-" * 50)

for col in numeric_features:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    
    # 상한선과 하한선 계산 (일반적으로 1.5배수 적용)
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    # 상한선 또는 하한선을 벗어나는 조건 필터링
    outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)][col]
    
    print(f"[{col}]")
    print(f"  - IQR 범위: {Q1} ~ {Q3} (IQR = {IQR:.2f})")
    print(f"  - 정상 범위 기준: {lower_bound:.2f} ~ {upper_bound:.2f}")
    print(f"  - 탐지된 이상치 개수: {len(outliers)}개")
    if len(outliers) > 0:
        print(f"  - 실제 이상치 값들: {outliers.values}")
    print("-" * 50)

    # 이상치 수학적 탐지(IQR)

=== IQR 기반 수치형 특성별 이상치 탐지 ===
--------------------------------------------------
[age]
  - IQR 범위: 48.0 ~ 61.0 (IQR = 13.00)
  - 정상 범위 기준: 28.50 ~ 80.50
  - 탐지된 이상치 개수: 0개
--------------------------------------------------
[trestbps]
  - IQR 범위: 120.0 ~ 140.0 (IQR = 20.00)
  - 정상 범위 기준: 90.00 ~ 170.00
  - 탐지된 이상치 개수: 9개
  - 실제 이상치 값들: [172 180 200 174 178 192 180 178 180]
--------------------------------------------------
[chol]
  - IQR 범위: 211.0 ~ 275.0 (IQR = 64.00)
  - 정상 범위 기준: 115.00 ~ 371.00
  - 탐지된 이상치 개수: 5개
  - 실제 이상치 값들: [417 407 564 394 409]
--------------------------------------------------
[thalach]
  - IQR 범위: 133.5 ~ 166.0 (IQR = 32.50)
  - 정상 범위 기준: 84.75 ~ 214.75
  - 탐지된 이상치 개수: 1개
  - 실제 이상치 값들: [71]
--------------------------------------------------
[oldpeak]
  - IQR 범위: 0.0 ~ 1.6 (IQR = 1.60)
  - 정상 범위 기준: -2.40 ~ 4.00
  - 탐지된 이상치 개수: 5개
  - 실제 이상치 값들: [6.2 5.6 4.2 4.2 4.4]
--------------------------------------------------


In [12]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import RobustScaler, OneHotEncoder

def clean_and_prepare_data(df_input):
    """
    파이프라인 진입 전, 중복 제거 및 불필요한 컬럼을 정제하는 함수
    """
    df_clean = df_input.copy()
    
    # 1. 중복 레코드 제거
    duplicated_count = df_clean.duplicated().sum()
    if duplicated_count > 0:
        print(f"[정제] 중복 데이터 {duplicated_count}건을 제거합니다.")
        df_clean = df_clean.drop_duplicates()
        
    # 2. 완전히 비어있는(모든 값이 결측치인) 컬럼 제거
    empty_cols = [col for col in df_clean.columns if df_clean[col].isnull().all()]
    if empty_cols:
        print(f"[정제] 완전히 비어있는 컬럼 제거: {empty_cols}")
        df_clean = df_clean.drop(columns=empty_cols)
        
    return df_clean

# --- [단계 1] 원본 데이터 병합 및 정제 ---
X = heart_disease.data.features 
y = heart_disease.data.targets 
raw_df = pd.concat([X, y], axis=1)

# 타겟 이진 분류화 및 데이터 정제
raw_df['target'] = raw_df['num'].apply(lambda x: 1 if x > 0 else 0)
raw_df = raw_df.drop(columns=['num'])

cleaned_df = clean_and_prepare_data(raw_df)

# --- [단계 2] Feature와 Target 분리 및 Train/Test 분할 ---
X_clean = cleaned_df.drop(columns=['target'])
y_clean = cleaned_df['target']

# 클래스 비율을 유지하며 분할 (stratify 적용)
X_train, X_test, y_train, y_test = train_test_split(
    X_clean, y_clean, test_size=0.2, random_state=42, stratify=y_clean
)

# --- [단계 3] 데이터 타입별 타겟 컬럼 정의 ---
# 앞서 분석한 기술 통계량을 바탕으로 정의합니다.
numeric_features = ['age', 'trestbps', 'chol', 'thalach', 'oldpeak']
categorical_features = ['sex', 'cp', 'fbs', 'restecg', 'exang', 'slope', 'ca', 'thal']

# --- [단계 4] 컴포넌트별 파이프라인 구성 (sklearn.pipeline.Pipeline) ---
# 1) 연속형 변수 파이프라인: 중앙값 대치 -> Robust 스케일러 적용
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', RobustScaler())
])

# 2) 범주형 변수 파이프라인: 최빈값 대치 -> 원핫 인코딩(학습 때 없던 카테고리는 무시)
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

# 3) ColumnTransformer를 이용해 컬럼별로 파이프라인을 매핑
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)
    ]
)

# --- [단계 5] 파이프라인 실행 및 데이터 누수 검증 ---
# Train 데이터로만 기준을 잡습니다 (fit_transform)
X_train_transformed = preprocessor.fit_transform(X_train)

# Test 데이터는 이미 학습된 기준을 '적용'만 합니다 (transform) - 데이터 누수 원천 차단!
X_test_transformed = preprocessor.transform(X_test)

print("\n=== 파이프라인 구성 및 변환 완료 ===")
print(f"변환된 훈련 데이터 크기(Shape): {X_train_transformed.shape}")
print(f"변환된 테스트 데이터 크기(Shape): {X_test_transformed.shape}")

# Scikit-Learn 기반 프로덕션 파이프라인 구현


=== 파이프라인 구성 및 변환 완료 ===
변환된 훈련 데이터 크기(Shape): (242, 28)
변환된 테스트 데이터 크기(Shape): (61, 28)


# EDA가 무엇을 알려 주었으며, 그에 따라 전처리에서 무엇을 어떻게 바꾸었는가?
**EDA를 통해 발견한 데이터 특징**

1. 타겟 클래스의 비율 (Class Distribution)
   - 심장병 환자(1)와 정상군(0)의 비율이 약 46% 대 54%로 확인되었습니다. 완전한 극단적 불균형은 아니지만 약간의 불균형이 존재합니다. 모델 평가 시 단순 '정확도(Accuracy)'만 보면 성능이 과대평가될 위험이 있습니다.

2. 연속형 변수의 이상치 (Outliers in Numeric Features)
   - Boxplot과 IQR 탐지 결과, `chol`(콜레스테롤)과 `trestbps`(휴식기 혈압) 등에서 정상 범위를 크게 벗어나는 극단적인 고수치 이상치들이 발견되었습니다.이 이상치들은 제거해야 할 오류 데이터가 아니라, 실제 심장병 고위험군 환자의 중요한 임상 신호(Signal)입니다.

3. 결측치의 존재 (Missing Values)
   - 수치형 변수에는 결측치가 없으나, 범주형/정수형 특성인 `ca`와 `thal` 변수에서 일부 결측치가 존재함을 확인했습니다.

---

### 데이터 특징에 따른 전처리 파이프라인 설계

위의 EDA 인사이트를 바탕으로, 새로운 데이터에도 안전하게 재적용 가능한 `scikit-learn Pipeline` 기반의 프로덕션 파이프라인을 아래와 같이 설계했습니다.

삭제 대신 대치(Imputation) 전략을 선택하였으며, 이상치 보존 및 `RobustScaler` 채택하고, 데이터 누수(Data Leakage)를 원천 차단하였습니다.